### PR TITLE
AVM: fix early eval exits for Debugger

### PR DESCRIPTION
**Pull request by jannotti**
When the Debugger is set, we were using a defer() to set the debugstate on every exit from eval().  Unfortunately, we exited for some conditions before the debugState was setup, so there was a nil pointer dereference while trying to communicate the error to the Debugger.

This moves those error checks down until after the debugState is setup. It also has the tiny advantage of not setting up the defer() unless it is needed (cx.Debugger is set)

(This contains some final specs changes approved by Foundation as well.)

### CODE REVIEW

File: data/transactions/logic/README_in.md [hunk 1]
```diff
@@ -30,8 +30,15 @@ arguments from it and pushing results to it. Some operations have
 _immediate_ arguments that are encoded directly into the instruction,
 rather than coming from the stack.
 
-The maximum stack depth is 1000. If the stack depth is
-exceeded or if a byte-array element exceed 4096 bytes, the program fails.
+The maximum stack depth is 1000. If the stack depth is exceeded or if
+a byte-array element exceed 4096 bytes, the program fails. If an
```
[the remark this answers is not in the document]
Comment (jannotti) [reply data/transactions/logic/README_in.md]: Sure, but I'm not going to go back and update the foundation spec.

[the remark this answers is not in the document]
Comment (jannotti) [reply data/transactions/logic/README_in.md]: And of course, now it's going to fail codegen test

[the remark this answers is not in the document]
Comment (michaeldiamant) [reply data/transactions/logic/README_in.md]: @jannotti Right - Sorry, let's disregard then.  It's too trivial to justify the effort involved.  I should have considered it before opening the thread.

[the remark this answers is not in the document]
Comment (jannotti) [reply data/transactions/logic/README_in.md]: Already committed b/c I had an itchy trigger finger. Will fix.

**Review by michaeldiamant — APPROVED** [review]
@jannotti Thanks for fixing - I confirmed `TestDryrunEarlyExit` fails in `master` as intended.

**Pull request merged**